### PR TITLE
Support kubernetes-anywhere provider

### PR DESCRIPTION
**Pull request by jamiehannaford**
**What this PR does / why we need it**:

Implements a new `kubernetes-anywhere` provider to allow upgrade testing in the e2e binary. This is the final step to allow https://github.com/kubernetes/test-infra/pull/4495 and https://github.com/kubernetes/kubernetes-anywhere/pull/450.

**Which issue this PR fixes**:

https://github.com/kubernetes/kubeadm/issues/311

**Special notes for your reviewer**:

Some questions I had

- Does the `--provider` flag specified [here](https://github.com/jamiehannaford/test-infra/blob/dbbf6261e0b44fc81944ed5c3b7152005cc0f1ff/jobs/config.json#L8587) get sent to the flag defined [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/test_context.go#L219)? Or should I add another `--provider` flag inside `--upgrade_args` like this: `--upgrade_args=... --provider=kubernetes-anywhere`?
- Is it necessary to add waiting logic after the `make` command, or will it implicitly handle that by itself?

Some other points:

- I chose `sed` to manipulate the current kubernetes-anywhere `.config` rather than duplicating another [`anywhere.go`](https://github.com/kubernetes/test-infra/blob/master/kubetest/anywhere.go). One suggestion was to use `jq` but since the config on disk is not serialized to JSON yet, I'm not sure how that'd work.
- Since I don't have a GCE/GKE account or vCenter, I can't actually verify the e2e binary works. I've managed to build it, but if somebody could quickly run a smoke test, I'd appreciate it. This is my first poke around test-infra and e2e, so there might be some plumbing missing

/cc @jessicaochen @luxas @pipejakob @roberthbailey

### CODE REVIEW

File: test/e2e/framework/nodes_util.go
Comment (roberthbailey) [inline test/e2e/framework/nodes_util.go]: should this at least log an error if it fails?

File: test/e2e/framework/nodes_util.go
Comment (roberthbailey) [inline test/e2e/framework/nodes_util.go]: if _, _, err := RunCmd(...); err != nil { return err }

File: test/e2e/framework/nodes_util.go
Comment (roberthbailey) [inline test/e2e/framework/nodes_util.go]: same here

File: test/e2e/framework/nodes_util.go
Comment (luxas) [inline test/e2e/framework/nodes_util.go]: Do we actually run deploy twice? Shouldn't we run `upgrade` or something like that to differ from net-new deployment?

**Review by luxas — COMMENTED** [review]
Did you get around to test this manually?

File: test/e2e/framework/nodes_util.go
Comment (jessicaochen) [reply test/e2e/framework/nodes_util.go]: This should be the upgrade-master command. Looks like the following on the command line:

make upgrade-master

**Comment by jessicaochen** [comment]
Does the --provider flag specified here get sent to the flag defined here? Or should I add another --provider flag inside --upgrade_args like this: --upgrade_args=... --provider=kubernetes-anywhere?
The config propagation is pretty confusing. @krousey do you know?

Is it necessary to add waiting logic after the make command, or will it implicitly handle that by itself?
The control plane should be upgraded when the command completes successfully. Should not need to wait.

**Comment by krousey** [comment]
@jessicaochen I think the logic you're looking for would reside in the deployer for kubernetes-anywhere in the kubetest binary. See https://github.com/kubernetes/test-infra/tree/master/kubetest. I don't know what the current implementation for that deployer/flag combination is, but that's probably where it would be defined and need to be changed. I wouldn't leave it to each individual job to pass that in the upgrade args if it's always needed.

**Comment by jamiehannaford** [comment]
@krousey We're already defining additional args [here](https://github.com/kubernetes/test-infra/pull/4495/files#diff-5539d275f091e62154f51653bbe4fba0R8324), which will then be passed by kubetest to the e2e binary [here](https://github.com/kubernetes/test-infra/blob/master/kubetest/e2e.go#L167) I assume (I'm not 100% familiar with k8s testing topology). 

My question was whether the `--provider` flag in `kubernetes/kubernetes/test/e2e/test_context` is set by passing in to `--upgrade_args`, or whether that job's `--provider` flag is somehow used instead. i.e. does the job look like this:

```
"--provider=kubernetes-anywhere",
"--upgrade_args=--ginkgo.focus=\\[Feature:MasterUpgrade\\] --ginkgo.focus=\\[Conformance\\] --upgrade-target=ci/latest-1.7"
```

or like this:

```
"--provider=kubernetes-anywhere",
"--upgrade_args=--ginkgo.focus=\\[Feature:MasterUpgrade\\] --ginkgo.focus=\\[Conformance\\] --upgrade-target=ci/latest-1.7 -provider=kubernetes-anywhere"
```

I assume it's the latter because looking at other jobs in config.json, they also provide flags in `--upgrade_args` which are defined in test_context.

**Review by luxas — APPROVED** [review]
Is this ok to merge now @roberthbailey @jessicaochen ?
LGTM

**Comment by krousey** [comment]
@jamiehannaford My suspicion is that you shouldn't have to pass the provider in the upgrade_args flag. kubetest is the piece that coordinates what flags go to which e2e invocation, which is why I suggested looking there. No other upgrade test has the provider defined in upgrade_args.

**Comment by jamiehannaford** [comment]
@krousey Okay, well we can try it out and submit another PR if things don't work.

**Review by luxas — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
@krousey @jessicaochen can you approve this?

**Comment by jamiehannaford** [comment]
@krousey @jessicaochen ping :)

**Comment by jessicaochen** [comment]
/assign @krousey

Out of all the approvers on the list, I think you would be the best due to your familiarity with upgrade testing code. Please review for approval. Thanks!

**Pull request merged**
